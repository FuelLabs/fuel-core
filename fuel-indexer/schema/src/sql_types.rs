#[cfg(feature = "db-models")]
use diesel::{
    backend::Backend,
    deserialize::{self, FromSql},
    pg::Pg,
    serialize::{self, IsNull, Output, ToSql},
};

#[cfg(feature = "db-models")]
use std::io::Write;

#[cfg(feature = "db-models")]
#[derive(SqlType)]
#[postgres(type_name = "Columntypename")]
pub struct Columntypename;

#[derive(Copy, Clone, Debug, PartialEq)]
#[cfg_attr(feature = "db-models", derive(AsExpression, FromSqlRow))]
#[cfg_attr(feature = "db-models", sql_type = "Columntypename")]
pub enum ColumnType {
    ID = 0,
    Address = 1,
    AssetId = 2,
    Bytes4 = 3,
    Bytes8 = 4,
    Bytes32 = 5,
    ContractId = 6,
    Salt = 7,
    Int4 = 8,
    Int8 = 9,
    UInt4 = 10,
    UInt8 = 11,
    Timestamp = 12,
    Blob = 13,
}

#[cfg(feature = "db-models")]
impl<DB: Backend> ToSql<Columntypename, DB> for ColumnType {
    fn to_sql<W: Write>(&self, out: &mut Output<W, DB>) -> serialize::Result {
        match *self {
            ColumnType::ID => out.write_all(b"ID")?,
            ColumnType::Address => out.write_all(b"Address")?,
            ColumnType::AssetId => out.write_all(b"AssetId")?,
            ColumnType::Bytes4 => out.write_all(b"Bytes4")?,
            ColumnType::Bytes8 => out.write_all(b"Bytes8")?,
            ColumnType::Bytes32 => out.write_all(b"Bytes32")?,
            ColumnType::ContractId => out.write_all(b"ContractId")?,
            ColumnType::Int4 => out.write_all(b"Int4")?,
            ColumnType::Int8 => out.write_all(b"Int8")?,
            ColumnType::UInt4 => out.write_all(b"UInt4")?,
            ColumnType::UInt8 => out.write_all(b"UInt8")?,
            ColumnType::Timestamp => out.write_all(b"Timestamp")?,
            ColumnType::Salt => out.write_all(b"Salt")?,
            ColumnType::Blob => out.write_all(b"Blob")?,
        }
        Ok(IsNull::No)
    }
}

#[cfg(feature = "db-models")]
impl FromSql<Columntypename, Pg> for ColumnType {
    fn from_sql(bytes: Option<&<Pg as Backend>::RawValue>) -> deserialize::Result<Self> {
        match not_none!(bytes) {
            b"ID" => Ok(ColumnType::ID),
            b"Address" => Ok(ColumnType::Address),
            b"AssetId" => Ok(ColumnType::AssetId),
            b"Bytes4" => Ok(ColumnType::Bytes4),
            b"Bytes8" => Ok(ColumnType::Bytes8),
            b"Bytes32" => Ok(ColumnType::Bytes32),
            b"ContractId" => Ok(ColumnType::ContractId),
            b"Int4" => Ok(ColumnType::Int4),
            b"Int8" => Ok(ColumnType::Int8),
            b"UInt4" => Ok(ColumnType::UInt4),
            b"UInt8" => Ok(ColumnType::UInt8),
            b"Timestamp" => Ok(ColumnType::Timestamp),
            b"Salt" => Ok(ColumnType::Salt),
            b"Blob" => Ok(ColumnType::Blob),
            _ => Err("Unrecognized enum variant".into()),
        }
    }
}

impl From<ColumnType> for i32 {
    fn from(typ: ColumnType) -> i32 {
        match typ {
            ColumnType::ID => 0,
            ColumnType::Address => 1,
            ColumnType::AssetId => 2,
            ColumnType::Bytes4 => 3,
            ColumnType::Bytes8 => 4,
            ColumnType::Bytes32 => 5,
            ColumnType::ContractId => 6,
            ColumnType::Salt => 7,
            ColumnType::Int4 => 8,
            ColumnType::Int8 => 9,
            ColumnType::UInt4 => 10,
            ColumnType::UInt8 => 11,
            ColumnType::Timestamp => 12,
            ColumnType::Blob => 13,
        }
    }
}

impl From<i32> for ColumnType {
    fn from(num: i32) -> ColumnType {
        match num {
            0 => ColumnType::ID,
            1 => ColumnType::Address,
            2 => ColumnType::AssetId,
            3 => ColumnType::Bytes4,
            4 => ColumnType::Bytes8,
            5 => ColumnType::Bytes32,
            6 => ColumnType::ContractId,
            7 => ColumnType::Salt,
            8 => ColumnType::Int4,
            9 => ColumnType::Int8,
            10 => ColumnType::UInt4,
            11 => ColumnType::UInt8,
            12 => ColumnType::Timestamp,
            13 => ColumnType::Blob,
            _ => panic!("Invalid column type!"),
        }
    }
}

impl From<&str> for ColumnType {
    fn from(name: &str) -> ColumnType {
        match name {
            "ID" => ColumnType::ID,
            "Address" => ColumnType::Address,
            "AssetId" => ColumnType::AssetId,
            "Bytes4" => ColumnType::Bytes4,
            "Bytes8" => ColumnType::Bytes8,
            "Bytes32" => ColumnType::Bytes32,
            "ContractId" => ColumnType::ContractId,
            "Salt" => ColumnType::Salt,
            "Int4" => ColumnType::Int4,
            "Int8" => ColumnType::Int8,
            "UInt4" => ColumnType::UInt4,
            "UInt8" => ColumnType::UInt8,
            "Timestamp" => ColumnType::Timestamp,
            "Blob`" => ColumnType::Blob,
            _ => panic!("Invalid column type! {}", name),
        }
    }
}
