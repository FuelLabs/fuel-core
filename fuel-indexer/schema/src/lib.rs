#![cfg_attr(not(feature = "use-std"), no_std)]
#[cfg(feature = "db-models")]
#[macro_use]
extern crate diesel;
extern crate alloc;
use alloc::vec::Vec;

use core::convert::{TryFrom, TryInto};
use serde::{Deserialize, Serialize};
// serde_scale for now, can look at other options if necessary.
use crate::sql_types::ColumnType;
use serde_scale;

#[cfg(feature = "use-std")]
use sha2::{Digest, Sha256};

#[cfg(feature = "use-std")]
pub const BASE_SCHEMA: &str = include_str!("./base.graphql");

pub mod sql_types;

#[cfg(feature = "db-models")]
pub mod db;

pub use fuel_tx::{Address, Bytes32, Bytes4, Bytes8, Color, ContractId, Salt};

pub type ID = u64;

pub fn serialize(obj: &impl Serialize) -> Vec<u8> {
    serde_scale::to_vec(obj).expect("Serialize failed")
}

pub fn deserialize<'a, T: Deserialize<'a>>(vec: &'a Vec<u8>) -> T {
    serde_scale::from_slice(&vec).expect("Deserialize failed")
}

#[cfg(feature = "use-std")]
pub fn type_id(type_name: &str) -> u64 {
    let mut bytes = [0u8; 8];
    bytes.copy_from_slice(&Sha256::digest(type_name.as_bytes())[..8]);
    u64::from_le_bytes(bytes)
}

#[cfg(feature = "use-std")]
pub fn schema_version(schema: &str) -> String {
    format!("{:x}", Sha256::digest(schema.as_bytes()))
}

#[derive(Debug, PartialEq, Deserialize, Serialize)]
pub enum FtColumn {
    ID(u64),
    Address(Address),
    Bytes4(Bytes4),
    Bytes8(Bytes8),
    Bytes32(Bytes32),
    Color(Color),
    ContractId(ContractId),
    Salt(Salt),
}

impl FtColumn {
    pub fn new(ty: ColumnType, size: usize, bytes: &[u8]) -> FtColumn {
        match ty {
            ColumnType::ID => {
                let ident =
                    u64::from_le_bytes(bytes[..size].try_into().expect("Invalid slice length"));
                FtColumn::ID(ident)
            }
            ColumnType::Address => {
                let address = Address::try_from(&bytes[..size]).expect("Invalid slice length");
                FtColumn::Address(address)
            }
            ColumnType::Bytes4 => {
                let bytes = Bytes4::try_from(&bytes[..size]).expect("Invalid slice length");
                FtColumn::Bytes4(bytes)
            }
            ColumnType::Bytes8 => {
                let bytes = Bytes8::try_from(&bytes[..size]).expect("Invalid slice length");
                FtColumn::Bytes8(bytes)
            }
            ColumnType::Bytes32 => {
                let bytes = Bytes32::try_from(&bytes[..size]).expect("Invalid slice length");
                FtColumn::Bytes32(bytes)
            }
            ColumnType::Color => {
                let color = Color::try_from(&bytes[..size]).expect("Invalid slice length");
                FtColumn::Color(color)
            }
            ColumnType::ContractId => {
                let contract_id =
                    ContractId::try_from(&bytes[..size]).expect("Invalid slice length");
                FtColumn::ContractId(contract_id)
            }
            ColumnType::Salt => {
                let salt = Salt::try_from(&bytes[..size]).expect("Invalid slice length");
                FtColumn::Salt(salt)
            }
            ColumnType::Blob => {
                panic!("Blob not supported for FtColumn!");
            }
        }
    }

    #[cfg(feature = "use-std")]
    pub fn query_fragment(&self) -> String {
        match self {
            FtColumn::ID(value) => {
                format!("{}", value)
            }
            FtColumn::Address(value) => {
                format!("'{:x}'", value)
            }
            FtColumn::Bytes4(value) => {
                format!("'{:x}'", value)
            }
            FtColumn::Bytes8(value) => {
                format!("'{:x}'", value)
            }
            FtColumn::Bytes32(value) => {
                format!("'{:x}'", value)
            }
            FtColumn::Color(value) => {
                format!("'{:x}'", value)
            }
            FtColumn::ContractId(value) => {
                format!("'{:x}'", value)
            }
            FtColumn::Salt(value) => {
                format!("'{:x}'", value)
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[cfg_attr(feature = "use-std", test)]
    fn test_fragments() {
        let id = FtColumn::ID(123456);
        let addr = FtColumn::Address(Address::try_from([0x12; 32]).expect("Bad bytes"));
        let bytes4 = FtColumn::Bytes4(Bytes4::try_from([0xF0; 4]).expect("Bad bytes"));
        let bytes8 = FtColumn::Bytes8(Bytes8::try_from([0x9D; 8]).expect("Bad bytes"));
        let bytes32 = FtColumn::Bytes32(Bytes32::try_from([0xEE; 32]).expect("Bad bytes"));
        let color = FtColumn::Color(Color::try_from([0xA5; 32]).expect("Bad bytes"));
        let contractid = FtColumn::ContractId(ContractId::try_from([0x78; 32]).expect("Bad bytes"));
        let salt = FtColumn::Salt(Salt::try_from([0x31; 32]).expect("Bad bytes"));

        insta::assert_yaml_snapshot!(id.query_fragment());
        insta::assert_yaml_snapshot!(addr.query_fragment());
        insta::assert_yaml_snapshot!(bytes4.query_fragment());
        insta::assert_yaml_snapshot!(bytes8.query_fragment());
        insta::assert_yaml_snapshot!(bytes32.query_fragment());
        insta::assert_yaml_snapshot!(color.query_fragment());
        insta::assert_yaml_snapshot!(contractid.query_fragment());
        insta::assert_yaml_snapshot!(salt.query_fragment());
    }
}
