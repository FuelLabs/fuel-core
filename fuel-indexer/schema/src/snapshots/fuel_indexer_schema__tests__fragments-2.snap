---
source: fuel-indexer/schema/src/lib.rs
expression: addr.query_fragment()

---
"'1212121212121212121212121212121212121212121212121212121212121212'"

